### ISSUE TITLE
Metric Depth ONNX

### ISSUE THREAD

**Issue by dnflslwlq** [fabio-sim/Depth-Anything-ONNX#16]
Hello, @fabio-sim 

I am trying to use the metric depth estimation pretrained model (vkitti) from Depth Anything V2 to convert it to ONNX and perform inference. (test : kitti dataset)

The relative depth did not show significant performance degradation. However, when converting the metric depth to ONNX and performing inference, the performance degradation was very severe.

I tried converting it with Dynamo and using a general script, but the result was the same.

What could be the cause of this?

The backbone model structures for relative and metric are not different, so why does the performance significantly degrade when converting to ONNX? I conducted the conversion using FP32. (not FP16)

The following results are from relative ONNX and metric ONNX, respectively.

![0000000000](https://github.com/fabio-sim/Depth-Anything-ONNX/assets/114914559/85d138f5-ff8e-4bf5-9d0d-b7cb6c244eb4)

![0000000000](https://github.com/fabio-sim/Depth-Anything-ONNX/assets/114914559/0cbccf47-f53f-4a3b-8f4b-0dc69cf0afff)

**Comment by fabio-sim** [comment]
Hi @dnflslwlq, thank you for your interest in Depth-Anything-ONNX.

Usually I observe a rtol difference of 1e-4 to 1e-3 between PyTorch & ONNXRuntime results, which is due to differing implementations of the underlying operations.

However, I'm not sure how the degradation could become that large. I added support for metric depth models in 
3128cb9 as well as the corresponding models in https://github.com/fabio-sim/Depth-Anything-ONNX/releases/tag/v2.0.0. It could be that there's a difference that I haven't taken into account..

**Comment by dnflslwlq** [comment]
@fabio-sim 
Thank you for your response.

The solution I found was to set the dummy input to a multiple of 14 that is closest to the test data resolution, instead of 518x518, when converting to ONNX. By doing so, the degradation was almost negligible.

Thank you for your interest and efforts in solving this issue together.

**Comment by PizzaRoll04** [comment]
Hey @dnflslwlq would you be willing to share the scripts and commands you used to convert to onnx?

Thanks!

**Comment by dnflslwlq** [comment]
Hi, @PizzaRoll04 
Please refer to `depth_anything_v2/export_v2.py` in this repository. https://github.com/spacewalk01/depth-anything-tensorrt/blob/main/depth_anything_v2/export_v2.py

here is my edited script : 
step 1) edit `export_v2.py`
```
import argparse
import os
import torch
import torch.onnx
from depth_anything_v2.dpt import DepthAnythingV2

def main():
    parser = argparse.ArgumentParser(description='Depth Anything V2')
    
    parser.add_argument('--height', type=int, default=518, help='Input height for the model')
    parser.add_argument('--width', type=int, default=518, help='Input width for the model')
    parser.add_argument('--encoder', type=str, default='vitl', choices=['vits', 'vitb', 'vitl', 'vitg'])
    parser.add_argument('--output', type=str, required=True, help='Output path for the ONNX model')

    args = parser.parse_args()
    
    model_configs = {
        'vits': {'encoder': 'vits', 'features': 64, 'out_channels': [48, 96, 192, 384]},
        'vitb': {'encoder': 'vitb', 'features': 128, 'out_channels': [96, 192, 384, 768]},
        'vitl': {'encoder': 'vitl', 'features': 256, 'out_channels': [256, 512, 1024, 1024]},
        'vitg': {'encoder': 'vitg', 'features': 384, 'out_channels': [1536, 1536, 1536, 1536]}
    }
    
    dataset = 'vkitti'
    max_depth = 80
    
    # model 
    depth_anything = DepthAnythingV2(**{**model_configs[args.encoder], 'max_depth': max_depth})
    depth_anything.load_state_dict(torch.load(f'checkpoints/depth_anything_v2_metric_{dataset}_{args.encoder}.pth', map_location='cpu'))
    depth_anything = depth_anything.to('cpu').eval()   

    # Define dummy input data with the specified height and width
    dummy_input = torch.ones((1, 3, args.height, args.width))

    # Provide an example input to the model, this is necessary for exporting to ONNX
    example_output = depth_anything.forward(dummy_input)
    
    onnx_path = args.output
    
    # Ensure the directory for onnx_path exists
    os.makedirs(os.path.dirname(onnx_path), exist_ok=True)
    
    # Export the PyTorch model to ONNX format
    torch.onnx.export(depth_anything, dummy_input, onnx_path, opset_version=19, input_names=["input"], output_names=["output"], verbose=False)
    
    print(f"Model exported to {onnx_path}")

if __name__ == "__main__":
    main()


```
step2) follow this manual. 
https://github.com/spacewalk01/depth-anything-tensorrt?tab=readme-ov-file#depth-anything-v2

but if you want to export metric depth onnx, you should copy `dpt.py` paste to `<depth_anything_installpath>/metric_depth/depth_anything_v2`
and also copy export_v2.py to `<depth_anything_installpath>/metric_depth`

and run this command line in `Depth-Anything-V2/metric_depth` directory: 
```
python export_v2.py --height 378 --width 1246 --encoder {one of vits, vitb, vitl} --output {your own output directory}
```

In this line, `--height` and `--width` should be multiples of 14 and the closest possible size to your input image (though latency is not guaranteed). For instance, 378x1246 is a multiple of 14 and is the closest size for the KITTI dataset input(375 1242)

but this is only for outdoor datasets. (The checkpoints files are pretrained model by vKITTI datasets)
if you want to infer indoor, edit this part 
```
    dataset = 'hypersim'
    max_depth = 20
```

**Comment by PizzaRoll04** [comment]
@dnflslwlq Thank you for your reply! 

For inference do you just use dynamo.py or do you do your own preprocessing, inference, and post-processing in some other script?

**Comment by dnflslwlq** [comment]
@PizzaRoll04 
I have written another inference script, but the overall flow of executing the ONNX session is very similar to `dynamo.py`.

For preprocessing, I use the method from `dynamo.py`. As you might know from referring to the `dpt.py` part of `Depth-Anything-V2`, this script performs preprocessing in the same way.

The post-processing method is different:

1. Instead of using `depth.max()` and `depth.min()` for normalization, I map the inferred meter values to pixel values between 0 and 80. This is because, being metric depth, the distance from the camera is important.
: `depth = (depth - 0) / (80 - 0) * 255.0  # max depth : 80`
2. I use the `Spectral colormap` from matplotlib's `colormaps.get_cmap`. This is the same method used in `depth-anything-v2`.

**Comment by PizzaRoll04** [comment]
@dnflslwlq Thank you for your help!

**Comment by ynma-hanvo** [comment]
hi,  I have the same problem with metric = indoor.   i got all depth = 0 , however the metric=none works fine. 
according to your suggestion, i need to change export resolution , do you know it for the hypersim dataset

**Comment by dnflslwlq** [comment]
@ynma-hanvo 
Hello, it seems that your case is different from mine. My issue is that when I run model inference, after preprocessing the input by changing the resolution to the expected value (default: 518x518), the output resolution is restored to the original image size (which is quite different from 518x518), leading to a degradation in quality. This is different from the problem of all depth values being 0.

However, according to the author of Depth Anything v2, the maximum depth is set to 80 for outdoor settings and 20 for indoor settings.

It might be that the depth values you inferred are not within the range between the minimum and maximum values, which could be causing the issue.

But there's something you mentioned that I don’t quite understand. If i assume that `metric=indoor` was used with the pretrained model of the Hypersim dataset, what model did you use with `metric=none`?

Could you describe your problem in more detail?

**Comment by deanofthewebb** [comment]
> dynamo

@ynma-hanvo I am also getting all 0s from the metric onnx file. I can run the relative depth onnx file fine but I cannot get any outputs with the metric file, this is my script, any help would be appreciated!

```
import cv2
from PIL import Image
import onnxruntime as ort
import numpy as np
import open3d as o3d
import time
import argparse
from pathlib import Path
import matplotlib
import os
import plotly.graph_objects as go
from transform import Resize, NormalizeImage, PrepareForNet
from torchvision.transforms import Compose



ppx = 8.709002501455901 * 10**2 # Principle point x in pixel.
ppy = 5.878986862144450 * 10**2 # Principle point y in pixel.
# focal_length_x_pixel = 1.017288703871180 * 10**4 # Focal length in pixel.
# focal_length_y_pixel = 1.018932963024356 * 10**4 # Focal length in pixel.
focal_length_x_pixel = 470.4 # 60
focal_length_y_pixel = 470.4 # 60
project_path = os.path.dirname(os.path.realpath(__file__))
project_path = Path(project_path)

def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "--img",
        type=str, 
        required=True,
        help="Path to input image.",
    )
    parser.add_argument(
        "--model",
        type=str,
        required=True,
        help="Path to ONNX model.",
    )
    return parser.parse_args()

def letterbox(img, target_shape=(518,518)):
    img_h, img_w, _ = img.shape
    new_h, new_w = target_shape[0], target_shape[1]
    offset_h, offset_w = 0, 0
    if (new_w / img_w) <= (new_h / img_h):
        new_h = int(img_h * new_w / img_w)
        offset_h = (target_shape[0] - new_h) // 2
    else:
        new_w = int(img_w * new_h / img_h)
        offset_w = (target_shape[1] - new_w) // 2
    resized = cv2.resize(img, (new_w, new_h),
                          interpolation=cv2.INTER_CUBIC)
    img = np.full((target_shape[0], target_shape[1], 3),
                  0,
                  dtype=np.uint8)
    img[offset_h:(offset_h + new_h),
        offset_w:(offset_w + new_w), :] = resized
    return img


def draw_geometries(geometries):
    graph_objects = []

    for geometry in geometries:
        geometry_type = geometry.get_geometry_type()

        if geometry_type == o3d.geometry.Geometry.Type.PointCloud:
            points = np.asarray(geometry.points)
            colors = None
            if geometry.has_colors():
                colors = np.asarray(geometry.colors)
            elif geometry.has_normals():
                colors = (0.5, 0.5, 0.5) + np.asarray(geometry.normals) * 0.5
            else:
                geometry.paint_uniform_color((1.0, 0.0, 0.0))
                colors = np.asarray(geometry.colors)

            scatter_3d = go.Scatter3d(x=points[:,0], y=points[:,1], z=points[:,2], mode='markers', marker=dict(size=1, color=colors))
            graph_objects.append(scatter_3d)

        if geometry_type == o3d.geometry.Geometry.Type.TriangleMesh:
            triangles = np.asarray(geometry.triangles)
            vertices = np.asarray(geometry.vertices)
            colors = None
            if geometry.has_triangle_normals():
                colors = (0.5, 0.5, 0.5) + np.asarray(geometry.triangle_normals) * 0.5
                colors = tuple(map(tuple, colors))
            else:
                colors = (1.0, 0.0, 0.0)

            mesh_3d = go.Mesh3d(x=vertices[:,0], y=vertices[:,1], z=vertices[:,2], i=triangles[:,0], j=triangles[:,1], k=triangles[:,2], facecolor=colors, opacity=0.60)
            graph_objects.append(mesh_3d)

    fig = go.Figure(
        data=graph_objects,
        layout=dict(
            scene=dict(
                xaxis=dict(visible=False),
                yaxis=dict(visible=False),
                zaxis=dict(visible=False)
            )
        )
    )
    fig.show()

def preprocess(raw_image, input_size=(1246, 378)):        
        transform = Compose([
            Resize(
                width=input_size[0],
                height=input_size[1],
                resize_target=False,
                keep_aspect_ratio=False,
                ensure_multiple_of=14,
                resize_method='lower_bound',
                image_interpolation_method=cv2.INTER_CUBIC,
            ),
            NormalizeImage(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
            PrepareForNet(),
        ])
        
        h, w = raw_image.shape[:2]
        
        image = cv2.cvtColor(raw_image, cv2.COLOR_BGR2RGB) / 255.0
        
        image = transform({'image': image})['image']
        image = np.expand_dims(image, axis=0)
        
        return image, (h, w)

def infer(args):
    """
    Depth-Anything V2 inference using ONNXRuntime. No dependency on PyTorch.
    """
    
    image_path = Path(args.img)
    model_path = project_path / 'weights' / args.model
    model_width, model_height = 518, 518

    # # Preprocessing, implement this part in your chosen language:
    # raw_image = cv2.imread(image_path.as_posix())
    # input, (h, w) = preprocess(raw_image, input_size=(model_width, model_height))
    # print('input', input.shape)

    image = cv2.imread(str(image_path))
    h, w = image.shape[:2]
    image = cv2.cvtColor(image, cv2.COLOR_BGR2RGB) / 255.0
    image = cv2.resize(image, (model_width, model_height), interpolation=cv2.INTER_CUBIC)
    image = (image - [0.485, 0.456, 0.406]) / [0.229, 0.224, 0.225]
    input = image.transpose(2, 0, 1)[None].astype("float32")

    color_image = Image.open(image_path).convert('RGB')

    # Inference
    sess_options = ort.SessionOptions()
    sess_options.enable_profiling = False

    providers = ["CPUExecutionProvider"]

    session = ort.InferenceSession(
        model_path, sess_options=sess_options, providers=providers
    )
    binding = session.io_binding()
    ort_input = session.get_inputs()[0].name
    binding.bind_cpu_input(ort_input, input)
    ort_output = session.get_outputs()[0].name
    binding.bind_output(ort_output, 'cpu')

    session.run_with_iobinding(binding)  # Actual inference happens here.
    depth = binding.get_outputs()[0].numpy()[0]

    cmap = matplotlib.colormaps.get_cmap('Spectral')

    # Postprocessing, implement this part in your chosen language:
    depth = (depth - depth.min()) / (depth.max() - depth.min()) * 255.0
    depth = depth.astype("uint8")
    resized_depth = Image.fromarray(depth).resize((w, h), Image.NEAREST)
    depth = (cmap(depth)[:, :, :3] * 255)[:, :, ::-1].astype(np.uint8)
    depth = cv2.resize(depth, (w, h), interpolation=cv2.INTER_CUBIC)
    cv2.imwrite('depthanything.png', depth)
    

    # '''
    # Post Processing
    # Generate mesh grid and calculate point cloud coordinates
    # '''
    t0 = time.time()
    x, y = np.meshgrid(np.arange(w), np.arange(h))
    x = (x - model_width / 2) / focal_length_x_pixel
    y = (y - model_height / 2) / focal_length_y_pixel
    z = np.array(resized_depth)
    points = np.stack((np.multiply(x, z), np.multiply(y, z), z), axis=-1).reshape(-1, 3)
    print(f'Post Processing: {time.time() - t0} seconds')

    '''
    Create the point cloud and save it to the output directory
    '''
    colors = np.array(color_image).reshape(-1, 3) / 255.0
    pcd = o3d.geometry.PointCloud()
    pcd.points = o3d.utility.Vector3dVector(points)
    pcd.colors = o3d.utility.Vector3dVector(colors)

    o3d.visualization.draw_geometries([pcd])

if __name__ == "__main__":
    args = parse_args()
    infer(args)
```

**Comment by ynma-hanvo** [comment]
@deanofthewebb  Hi,  I used the dynamo.py file to do the inference for the exported metric = NONE model,  and it works.  
you can refer to the infer function.

**Comment by ynma-hanvo** [comment]
> @ynma-hanvo Hello, it seems that your case is different from mine. My issue is that when I run model inference, after preprocessing the input by changing the resolution to the expected value (default: 518x518), the output resolution is restored to the original image size (which is quite different from 518x518), leading to a degradation in quality. This is different from the problem of all depth values being 0.
> 
> However, according to the author of Depth Anything v2, the maximum depth is set to 80 for outdoor settings and 20 for indoor settings.
> 
> It might be that the depth values you inferred are not within the range between the minimum and maximum values, which could be causing the issue.
> 
> But there's something you mentioned that I don’t quite understand. If i assume that `metric=indoor` was used with the pretrained model of the Hypersim dataset, what model did you use with `metric=none`?
> 
> Could you describe your problem in more detail?

hi, metric=None use another pretrained model.    
and metric=Indoor use the depth_anything_v2_metric_hypersim_vits.pth  pretrained model; 
hope I am correct.

**Comment by deanofthewebb** [comment]
> ne use another pretrained model.

@ynma-hanvo @dnflslwlq 

Were you able to get an output for the onnx model generated with metric=indoor or metric=outdoor? All of the metric onnx models I generated gave me all 0s as output so I just used the non-metric onnx files. However I now have to go to the Kitti Dataset and compare the ground truth values with the onnx predictions and hopefully determine some depth scaling factor heuristic to apply to the depth-anything v2 onnx predicted depths. My hope is that by getting the metric depth predictions to work in onnx, I can bypass this step.
